js/ui/messageTray.js

```diff
@@ -50,12 +50,15 @@ var State = {
 // These reasons are useful when we destroy the notifications received through
 // the notification daemon. We use EXPIRED for notifications that we dismiss
 // and the user did not interact with, DISMISSED for all other notifications
-// that were destroyed as a result of a user action, and SOURCE_CLOSED for the
-// notifications that were requested to be destroyed by the associated source.
+// that were destroyed as a result of a user action, SOURCE_CLOSED for the
+// notifications that were requested to be destroyed by the associated source,
+// and REPLACED for notifications that were destroyed as a consequence of a
+// newer version having replaced them.
 var NotificationDestroyedReason = {
     EXPIRED: 1,
     DISMISSED: 2,
-    SOURCE_CLOSED: 3
+    SOURCE_CLOSED: 3,
+    REPLACED: 4
```

Comment:
I understand you want to make the code as simple as possible, but IMO this is bit less safe for the purpose of the Shell Quest View, because it is not really a notification, so if we're relying on the original GTK behavior, we always have to make sure that changes from upstream won't break the Quest View.
Of course with my original changes, things could be affected in a rebase as well, but it was less dependent on the GTK notificatoins side.

Reply:
I consider this patch to be a needed improvement upstream regardless of the use we make of it in this PR.

I don't agree that this is less safe than relying on other behavior from a class not specifically written for this purpose. We'll have to make sure this feature keeps working one way or another after a rebase, so saying that relying on a specific behavior could be safer than relying on another is really a speculation right now. Either way, I will assume that we are OK proceeding with this change right now.

Reply:
Yeah. I think both views have their pros and cons. But I agree with moving on with what we have. Thanks.